### PR TITLE
[Fix #138] Allow the users to user the new GitHub's PR review feature

### LINKED ISSUE #138: Allow the users to use the new GitHub's PR review feature
Allow the projects to send a whole PR review (including all the comments for the issues found within the checked code). Something like this https://developer.github.com/v3/pulls/reviews/#create-a-pull-request-review

The user must be able to select which review style to use, either `pr_review` (feature added when this issue is fixed) or `individual_comments` (current working mechanism). `pr_review` will be the default one from now on.

Example: `application:get_env(egithub, review_style)`

### PR DESCRIPTION
**Pull request by harenson**
More info about GitHub's [PR Reviews](https://developer.github.com/v3/pulls/reviews/)

### CODE REVIEW

File: src/egithub_webhook.erl [hunk 1]
```diff
@@ -275,3 +279,52 @@ line_comment_exists(Comments, Path, Position, Body) ->
...
+          <<"state">> := State}) ->
+        case State of
+          <<"DISMISSED">> ->
+            % Don't try to dismiss an already dismissed review ;)
+            ok;
+          _ ->
+            Body = #{message => <<"Old/outdated review">>},
+            {ok, _} = egithub:dismiss_pr_review(Cred, Repo, PR, RId, Body)
+        end;
+       (_) ->
```
Comment (elbrujohalcon) [inline src/egithub_webhook.erl]: Indentation is weird here. It's particularly noticeable in the `(_) ->` branch of the case.
And why the parentheses here?

File: src/egithub_webhook.erl [hunk 2]
```diff
@@ -275,3 +279,52 @@ line_comment_exists(Comments, Path, Position, Body) ->
...
+        ok
+    end,
+  _ = lists:foreach(Fun, ExistingReviews),
+  % Without this sleep call, the dismissal message will appear below the
```
Comment (elbrujohalcon) [inline src/egithub_webhook.erl]: OK, but then the sleep call should be done _outside_ this function. This function only dismisses the PR review, it doesn't need to know that we're doing something else afterwards.

File: src/egithub_webhook.erl [hunk 3]
```diff
@@ -275,3 +279,52 @@ line_comment_exists(Comments, Path, Position, Body) ->
...
+          <<"state">> := State}) ->
+         case State of
+           <<"DISMISSED">> ->
+             % Don't try to dismiss an already dismissed review ;)
+             ok;
+           _ ->
+             Body = #{message => <<"Old/outdated review">>},
+             {ok, _} = egithub:dismiss_pr_review(Cred, Repo, PR, RId, Body)
+         end;
+       _ ->
```
Comment (gadgetci) [inline src/egithub_webhook.erl]: According to **Xref**:
> syntax error before: '->'

File: src/egithub_webhook.erl [hunk 4]
```diff
@@ -275,3 +279,52 @@ line_comment_exists(Comments, Path, Position, Body) ->
...
+      write_pr_review(Cred, Repo, PR, Messages)
+  end.
+
+-spec write_pr_review(egithub:credentials(), egithub:repository(), integer(),
+                      review())->
+  egithub:result().
+write_pr_review(Cred, Repo, PR, PrReview) ->
+  egithub:pr_review(Cred, Repo, PR, PrReview, #{post_method => queue}).
+
+-spec dismiss_old_pr_reviews(egithub:credentials(), egithub:repository(),
```
Comment (gadgetci) [inline src/egithub_webhook.erl]: According to **Xref**:
> spec for undefined function dismiss_old_pr_reviews/3

File: src/egithub_webhook.erl [hunk 5]
```diff
@@ -275,3 +279,52 @@ line_comment_exists(Comments, Path, Position, Body) ->
...
+handle_messages_or_review(Cred, Repo, PR, Messages) ->
+  {ok, ReviewStyle} = application:get_env(egithub, review_style, pr_review),
+  case ReviewStyle of
+    individual_comments ->
+      {ok, LineComments} = egithub:pull_req_comments(Cred, Repo, PR),
+      {ok, IssueComments} = egithub:issue_comments(Cred, Repo, PR),
+      Comments = LineComments ++ IssueComments,
+      write_comments(Cred, Repo, PR, Comments, Messages);
+    pr_review ->
+      dismiss_old_pr_reviews(Cred, Repo, PR),
```
Comment (gadgetci) [inline src/egithub_webhook.erl]: According to **Xref**:
> function dismiss_old_pr_reviews/3 undefined

**Comment by gadgetci** [comment]
**xref** failed with exit status: 1

**Pull request merged**